### PR TITLE
Allow governance controlled nomination pools 

### PR DESCRIPTION
**Pull request by kianenigma**
Fixes https://github.com/paritytech/substrate/issues/11479

This essentially allows a pool to be created, then its roles be destoryed, and henceforth it can only be controlled by the `root` origin. 

As noted in the issue, this is easier to implement, but slightly inflexible. 

The new `ensure_root_or_signed` function is a bit tricky to review, read with care.

Note that we cannot create a pool (trivially) with root, because the root cannot pay for anything. Although we can allow this as well, open to discussion. 

- [ ] Once #11426 is merged, the same functionality should be added for `chill` transaction as well.
- [ ] Re-calculate all weights.

### CODE REVIEW

File: frame/nomination-pools/src/lib.rs [hunk 1]
```diff
@@ -1799,67 +1802,117 @@ pub mod pallet {
...
+						ensure!(
+							bonded_pool.state != PoolState::Destroying &&
+								bonded_pool.can_toggle_state(&who),
+							Error::<T>::CanNotChangeState
+						);
+						bonded_pool.set_state(state);
+					}
+					Ok(bonded_pool)
+				},
+				|| {
```
Comment (rossbulat) [inline frame/nomination-pools/src/lib.rs:1863]: I'm noticing we are not checking `ok_to_be_open()` if root. Am I correct in assuming that governance's role here is to override these checks and force a state - such as re-opening a pool even if there are `OverflowRisk`s?

File: frame/nomination-pools/src/lib.rs [hunk 2]
```diff
@@ -1973,6 +2030,29 @@ pub mod pallet {
 }
 
 impl<T: Config> Pallet<T> {
+	/// Slightly more customizable version of [`frame_system::ensure_signed_or_root`].
```
Comment (rossbulat) [inline frame/nomination-pools/src/lib.rs:2033]: Pretty useful function, I can see it being useful in other pallets too if it were promoted to frame_system

**Review by rossbulat — COMMENTED** [review]
I notice the "holy" root terminology used here for the first time. Is this just another term for the root origin?

File: frame/nomination-pools/src/lib.rs [hunk 3]
```diff
@@ -1799,67 +1802,117 @@ pub mod pallet {
...
+			T::StakingInterface::nominate(bonded_account, validators.clone())?;
+			Self::deposit_event(Event::<T>::Nominated { pool_id, validators });
+
 			Ok(())
 		}
 
+		/// Set the metadata of a pool.
+		///
+		/// Dispatch origin of this call must be signed by the root or state-toggler roles of the
+		/// pool, or be the (holy) root origin.
```
Comment (shawntabrizi) [inline frame/nomination-pools/src/lib.rs:1835]: holy is cute, but "system root" or "frame root" will probably help clarity

File: frame/nomination-pools/src/lib.rs [hunk 4]
```diff
@@ -1799,67 +1802,117 @@ pub mod pallet {
...
 			metadata: Vec<u8>,
 		) -> DispatchResult {
-			let who = ensure_signed(origin)?;
+			Self::ensure_signed_or_root(
+				origin,
+				|who| {
+					ensure!(
+						BondedPool::<T>::get(pool_id)
+							.ok_or(Error::<T>::PoolNotFound)?
+							.can_set_metadata(&who),
```
Comment (shawntabrizi) [inline frame/nomination-pools/src/lib.rs:1895]: This is probably not necessary. Utility has `dispatch_as` which can emulate any user, so we just need a call to make sure that at least sets some user to be able to access functions like this, and then can just assume a signed origin.

Comment (kianenigma) [reply frame/nomination-pools/src/lib.rs:1895]: So, for instance, we can make this the procedure to have governance controlled pools: 

Set your roles all to a funded, key-less account (e.g. council account) and then the governance will then be able to dispatch all of these with `dispatch_as`. 

Other than needing that funded account to exist, it does seem like a nicer idea. That should not be an issue as we can have the treasury itself transfer some funds to a known-key-less account (`0x111..`). I like reusing `dispatch_as` here as well, it seem to have been made exactly for such use cases. 

If you agree with all of this, then PR can be closed.

Comment (shawntabrizi) [reply frame/nomination-pools/src/lib.rs:1895]: What I am saying is that there should just be a root call which can set the roles from NONE to some arbitrary account, callable by governance, and then that can be combined with `dispatch_as` to access and modify anything without any additional root checks.

Comment (kianenigma) [reply frame/nomination-pools/src/lib.rs:1895]: `update_roles` already supports root, so this is solved.

**Comment by kianenigma** [comment]
As noted in https://github.com/paritytech/substrate/pull/11496#discussion_r879029287, this feature is already possible using `dispatch_as`.

**Pull request closed without merging**